### PR TITLE
Refactor to use ES6 and Promises

### PR DESCRIPTION
**Pull request by andyhd**
* Refactored for readability

### CODE REVIEW

**Review by r4vi — APPROVED** [review]

**Comment by r4vi** [comment]
LGTM :+1:

File: rules/auth0-authorization-extension.js [hunk 1]
```diff
@@ -3,60 +3,44 @@
 *  Updated by kerin.cosford@digital.justice.gov.uk, 2017-01-14T22:47:58.651Z
  */
 function (user, context, callback) {
-  // If connection is not passwordless skip this rule
-  var connection = user.identities[0].connection;
+  const connection = user.identities[0].connection;
```
Comment (xoen) [inline rules/auth0-authorization-extension.js:6]: So we can use ES6 in rules? Are we 100% sure about that?

File: rules/auth0-authorization-extension.js [hunk 2]
```diff
@@ -3,60 +3,44 @@
...
+    .catch((err) => {
+      console.log(`Error from Authorization Extension: ${err.error}`);
+      callback(new UnauthorizedError(`Authorization Extension: ${err.message}`));
+    });
+  
+  function getUserPolicy() {
+    const rp = require('request-promise');
+    const endpoint = `users/${user.user_id}/policy/${context.clientID}`;
+    return rp({
+      method: 'POST',
```
Comment (xoen) [inline rules/auth0-authorization-extension.js:31]: I can't remember much about this rule but isn't weird that a `get...` function makes a `POST` HTTP request? As far as I can tell it's **get**ting the user policy off the Authz extension API, weird.

File: rules/auth0-authorization-extension.js (hunk 1 above)
Comment (andyhd) [reply rules/auth0-authorization-extension.js:6]: Yes

File: rules/auth0-authorization-extension.js (hunk 2 above)
Comment (andyhd) [reply rules/auth0-authorization-extension.js:31]: I know - it is a weird API.

File: rules/auth0-authorization-extension.js [hunk 3]
```diff
@@ -3,60 +3,44 @@
...
+      
+      callback(null, user, context);
+    })
+    .catch((err) => {
+      console.log(`Error from Authorization Extension: ${err.error}`);
+      callback(new UnauthorizedError(`Authorization Extension: ${err.message}`));
+    });
+  
+  function getUserPolicy() {
+    const rp = require('request-promise');
```
Comment (xoen) [inline rules/auth0-authorization-extension.js:28]: I see you've removed the check on the response status code and error, how does `request-promise` handle that?

Actually, I see that the error is handled in a `catch()` statement, does this library raise an error on non-200? Bear in mind we're posting, isn't the "normal" response status of an HTTP POST a [`201 CREATED`](https://httpstatusdogs.com/201-created)?

File: rules/auth0-authorization-extension.js [hunk 4]
```diff
@@ -3,60 +3,44 @@
...
-  // Get the policy for the user.
-  function getPolicy(user, context, cb) {
-    request.post({
-      url: EXTENSION_URL + "/api/users/" + user.user_id + "/policy/" + context.clientID,
+  
+  getUserPolicy()
+    .then((policy) => {
+      user.permissions = policy.permissions;
+      
+      if (!user.permissions.includes('view:app')) {
```
Comment (xoen) [inline rules/auth0-authorization-extension.js:16]: I can't remember if we actually use the `user.permissions` in the user object anywhere. Probably safer to keep at least for now.

**Review by xoen — APPROVED** [review]
Looks good 👍

File: rules/auth0-authorization-extension.js (hunk 3 above)
Comment (andyhd) [reply rules/auth0-authorization-extension.js:28]: The `simple: True` option to `rp()` causes it to throw an error on non-200 responses. I've tested the API, and it returns `200` for this request, not `201`.

**Pull request merged**